### DISCUSSION TITLE
Feedback - What I miss in Positron, what is really great (from a VSCode and former RStudio user)

### DISCUSSION THREAD

**Discussion by strengejacke** [posit-dev/positron#3686] [Feedback]
Short background: I used RStudio for quite a long time, then switched to VSCode about 1.5-2 years ago. I'm a package developer, so GitHub integration and package dev is important when I use an IDE. I only use R, so I don't care about Python, Julia, ...

I learned about Positron just yesterday, installed it and tried it out. And I'm really excited!

### What I like about Positron and missed in VSCode

- Context sensitive auto-completion is great. We have this in RStudio, but in VSCode, whenever I started typing a function, the suggested arguments were those from the default method. RStudio and Positron suggest arguments based on the input type, i.e. for the `.numeric()` or `.data.frame()` methods, when the input to a function is numeric or data frame.

- The built-in console is great! Much better than the default in VSCode, and I never felt happy with Radian, and really missed the RStudio console.

- Positron feels more "responsive". When searching for functions ("show all symbols"), it works pretty fast.

- R integration feels better in Positron than in VSCode.
- The `Testing` panel/feature is very nice! Even better than in RStudio, I think.

### What I like about Positron and missed in RStudio

Mostly those things that made me switch to VSCode

- GitHub integration (_way_ better than in RStudio)

- More flexibility regarding editor / window / pane layouts

- Extensions

- Modern look, including _much_ better colorization of bracket-pairs or color-codes.

## What do I miss in Positron?

Some things that I use often in VSCode are not yet working/available in Positron, or at least I haven't found them yet. Maybe it's because I'm using Positron since only a few hours now.

### Color theme

- Color theme: there a hardly any nice (dark) color themes. This is not a problem for Positron, it's the same for VSCode. But I really like the default VSCode color theme, `Dark+` (Default Dark+). Can we have that color theme as well? Or at least, fix some "issues":

  - In the Positron default theme, arguments like `fixed` have the same color as commands like `if` or `function` (argument names should have different colors than functions).
  - Package scopes (like `stats::`) has the same color like symbols/variable names (should be different).
  - In regression formulas, `y ~ x1 + x2`, all symbols have the same colors. I think that `+` etc. should have other colors than `x1` etc.
  - In `iris$Species`, the `$` should get a different color than the remaining names.
  - `if` is blue, but `ifelse` is yellow (as other functions). Not sure about this behaviour, maybe `if` and `ifelse` should get the same color?
  - Things like `,`, `<-`, `>` etc. have the same color as own variable names etc. - should be a different color.
  - When you have, e.g. `fixed = TRUE`, both `fixed` and `TRUE` get the same color. I think argument name and value should get different colors.

One advantage of the default _Positron_ color theme: user-defined functions have the same color as other functions, and thus can be better identified as functions.

**left: VSCode, right: Positron**

![image](https://github.com/posit-dev/positron/assets/26301769/0aefc65c-abe7-4c00-afda-4dcca462b8be)

### Diagnostics or lintr?

Especially for package development, the built-in diagnostic is too aggressive. It's normal that you have many internal functions, which are, however, all marked as "not in scope" when diagnostics are activated. I would like to _disable_ the diagnostics, but _enable_ lintr like in VSCode. Is that possible? I don't want the diagnostics, but I'd like to have lintrs like this one:

![image](https://github.com/posit-dev/positron/assets/26301769/14c846e5-a3dc-474a-ba77-a859ac117053)

### Co-Pilot

I think this is not necessary to say, but the GitHub Co-pilot extension is not available. This is really a missing feature!

### Viewer

I use the Viewer regularly to display HTML-tables. What I really _loved_ in RStudio, but miss in VSCode and Positron is a button in the viewer pane to open the currently displayed HTML-file in a browser.

### Debugging

If you come from VSCode, the debugging panel is a bit confusing, since you cannot install a debugger extension. I guess this feature is under development? Using `browser()` in the source helped, and the debug-panel looks great (like in VSCode). I suggest adding functionality to the debug panel (so it's less confusing how to use it) or at least clearly point out to using `browser()`.

### Key bindings for user-defined R command.

<strike>In VSCode, I had a keybinding to create a reprex:

```
    // Create reprex (must be in clipboard)
    {
      "description": "R Create Reprex",
      "key": "ctrl+shift+x",
      "command": "r.runCommand",
      "args": "reprex::reprex()",
      "when": "resourceLangId == 'r'"
    },
```

I used `"command": "r.runCommand"`. But `r.runCommand` is not available in Positron. Is this available through extensions?</strike>

Solution:

```
    // create reprex from selection
    {
        "key": "ctrl+shift+x",
        "command": "workbench.action.executeCode.console",
        "when": "editorTextFocus",
        "args": {
            "langId": "r",
            "code": "reprex::reprex_selection()",
            "focus": true
        }
    }
```
### Data viewer

The data viewer is nice, especially with the additional column with information about the variables. If possible, add variable labels (`label` attribute) to the view, below column names, like in RStudio.

### Editor

If possible, add a "Compile report" button, like in RStudio.

![image](https://github.com/posit-dev/positron/assets/26301769/b0dbb817-2725-4e53-8d26-692f612d845a)

## Summary

Positron looks really great, and I would love to switch! Would be very happy if you could address the above mentioned missing points.

**Comment by strengejacke** [comment]
Just saw the data viewer, pretty nice! If possible, add variable labels to the view.

Here's Positron:

![image](https://github.com/posit-dev/positron/assets/26301769/66e41d33-c929-4f21-891a-7d38f4d41951)

Here's RStudio (unfortunately, no information about type and missing values)

![image](https://github.com/posit-dev/positron/assets/26301769/7bb0f752-ca2c-4cc3-9d16-c3ca43b51b19)

**Reply by jthomasmock** [reply]
Glad you're liking the Data Explorer! We have a lot of improvements we'd like to make. If you want to, please upvote: https://github.com/posit-dev/positron/issues/2971

**Reply by DOH-FAA3303** [reply]
The data explorer is sweet but it seems to only work for R. Anyone else having trouble viewing Python dataframes? or maybe I need to configure something?

**Reply by jmcphers** [reply]
It should work for Python too! Can you tell us more about what you're seeing? (Error messages?)

**Reply by DOH-FAA3303** [reply]
@jmcphers I'm not sure if i can give you a reproducible example as there are no error messages. When I'm in R I am able to double click or select 'View' on a dataframe and the explorer will pop up. For Python nothing happens when I do either

![image](https://github.com/posit-dev/positron/assets/92396451/cb2afeed-9fc5-4879-b62c-9b5609ae73b6)

**Reply by DOH-FAA3303** [reply]
I spoke too soon. So, it doesn't appear to be working for python version 3.9.15. But when I try higher versions with jupyter installed into the env it seems to work.

**Reply by petetronic** [reply]
Thanks for sharing those details, our Polars support is still under development, but we'll take a look this week to see what might be happening for the version you mention.

**Reply by DOH-FAA3303** [reply]
@petetronic thanks! This isn't just with Polars though, it also happens with Pandas dataframes with Python version 3.9 and 3.10 (and possibly others).

**Comment by gadenbuie** [comment]
A great resource for finding your new favorite color theme is https://vscodethemes.com/. Not all themes are available in Positron (they need to publish the theme extension to https://open-vsx.org), but most are and there are some great ones to choose from.

**Reply by strengejacke** [reply]
I'm not that unsatisfied with the current theme, but rather that for some identifiers more distinct colors would be nice. That's of course a matter of taste, @lionel- just suggested the opposite: https://github.com/posit-dev/positron/issues/3688 ;-)

**Reply by strengejacke** [reply]
Will look at the themes, though, there are much more than you can see from inside Positron. Maybe I'll find something that I really like.

**Reply by gvelasq** [reply]
Two things I've missed in VS Code that I hope can make their way into Positron are the 'classic' RStudio themes and rsthemes.

The Tomorrow Night Bright theme as it's implemented by Microsoft [here](https://marketplace.visualstudio.com/items?itemName=ms-vscode.Theme-TomorrowKit) is very different across VS Code and RStudio. The use of red in the Microsoft version is quite distracting when writing R code. It would be great if an RStudio-style version of Tomorrow Night Bright could be ported to Positron.

**Reply by gvelasq** [reply]
Following up to add that I've ported RStudio's implementation of Tomorrow Night Bright to Positron. I am planning to publish it on the VS Code Marketplace and Open VSX.

Repository: https://github.com/gvelasq/tomorrow-night-bright-r-classic

Related discussion: https://github.com/posit-dev/positron/discussions/5842#discussioncomment-12048669

**Reply by gvelasq** [reply]
The 'Tomorrow Night Bright (R Classic)' extension is now published.

Open VSX: https://open-vsx.org/extension/gvelasq/tomorrow-night-bright-r-classic
VS Code Extension Marketplace: https://marketplace.visualstudio.com/items?itemName=gvelasq.tomorrow-night-bright-r-classic

**Comment by gadenbuie** [comment]
> I used `"command": "r.runCommand"`. But `r.runCommand` is not available in Positron. Is this available through extensions?

There's a more generic interface with two commands that can be used to execute code in any language: `workbench.action.executeCode.console` and `workbench.action.executeCode.console`. See details in #2684.

Here's an example shortcut you could use (add to the JSON file in the "Preference: Open Keyboard Shortcuts (JSON)" command):

```json
{
    "key": "Cmd+Shift+R",
    "command": "workbench.action.executeCode.console",
    "when": "editorTextFocus",
    "args": {
        "langId": "r",
        "code": "reprex::reprex_selection()",
        "focus": true
    }
}
```

**Reply by boshek** [reply]
@gadenbuie this is wonderful and should probably be added [here](https://github.com/posit-dev/positron/wiki/Keyboard-Shortcuts) (unless it documented elsewhere?). I'd do it myself but I don't think you can PR to a wiki. Thanks for this!

**Reply by gadenbuie** [reply]
@boshek that's a great idea! @jthomasmock or @juliasilge, could we add a short section to the [Keyboard Shortcuts wiki page](https://github.com/posit-dev/positron/wiki/Keyboard-Shortcuts) with instructions on adding your own keyboard shortcuts to call R or Python commands?

**Reply by juliasilge** [reply]
That's a great idea, yes! I will get this added in the next few days.

**Reply by juliasilge** [reply]
Got this added here: <https://github.com/posit-dev/positron/wiki/Keyboard-Shortcuts#custom-shortcuts>

**Comment by gadenbuie** [comment]
> I use the Viewer regularly to display HTML-tables. What I really _loved_ in RStudio, but miss in VSCode and Positron is a button in the viewer pane to open the currently displayed HTML-file in a browser.

There is a button for this in the Viewer pane, but not in the Plots pane. Shiny and Quarto app/document previews end up in the Viewer pane, but most htmlwidgets are displayed in the Plots pane. That may change to also send htmlwidgets to the Viewer pane, or Positron may decide to add the open in browser affordance to the Plots pane.

![image](https://github.com/posit-dev/positron/assets/5420529/dbad117a-68d0-4fcb-a622-09079e9ad6bb)

**Reply by strengejacke** [reply]
That's ok, I was only looking for the viewer pane right now. But how do you get that toolbar? I don't have it:

![image](https://github.com/posit-dev/positron/assets/26301769/c0c23bb2-cffa-4431-b2c7-72321ce8b669)

**Reply by gadenbuie** [reply]
Oh interesting. Something different happens for `gt` from other htmlwidgets like `leaflet`. I think currently that toolbar is hidden unless the file is served from a webserver. In this case, `gt` is likely being rendered to a static HTML file. I'll make sure this feedback makes it into the relevant issues.

**Comment by kyleGrealis** [comment]
![image](https://github.com/posit-dev/positron/assets/84798289/29596dfb-f11a-44c4-87d9-4ca59e99b49a)

An option you could explore to address the lintr issue is to add "00" (or another percentage -- "alpha" value) at the end of the hex codes for the warnings, errors, and info. Since they use RGBA, you're adjusting the alpha or transparency of those squiggly lines, but your file tab name color will still show red if there's existing errors. 

(~~Please ignore my first 3 color changes as those changes to the status bar don't quite work the same as in VSCode ... yet?~~)

**Reply by strengejacke** [reply]
Does this affect _diagnostics_ or a _lintr_ feature? I'm not sure if lintr support is already included or coming later. What I'd like to achieve is to disable diagnostic (they're still active after restart) and to enable lintr.

**Comment by lionel-** [comment]
Regarding R highlightings, some of this is tracked in https://github.com/posit-dev/positron/issues/3688

Regarding the R debugger, we are tracking this in https://github.com/posit-dev/positron/issues/3681

Regarding diagnostics, we have plans to drastically improve the situation. In the meantime you can disable them if they are too distracting, see the `positron.r.diagnostics.enable` setting.

**Reply by strengejacke** [reply]
Thanks! Disabling does not yet work (see https://github.com/posit-dev/positron/issues/3685)
